dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilder.java

```diff
@@ -318,46 +321,54 @@ private static void levelOrder(
               DependencyNode failedDependencyNode = artifactResult.getRequest().getDependencyNode();
               ExceptionAndPath failure =
                   ExceptionAndPath.create(parentNodes, failedDependencyNode, resolutionException);
-              if (isUnacceptableResolutionException(failure)) {
-                resolutionFailures.add(failure);
+              if (!isSafeResolutionException(failure)) {
+                unsafeResolutionFailures.add(failure);
               }
             }
           } catch (DependencyCollectionException collectionException) {
             DependencyNode failedDependencyNode = collectionException.getResult().getRoot();
             ExceptionAndPath failure =
                 ExceptionAndPath.create(parentNodes, failedDependencyNode, collectionException);
-            if (isUnacceptableResolutionException(failure)) {
-              resolutionFailures.add(failure);
+            if (!isSafeResolutionException(failure)) {
+              unsafeResolutionFailures.add(failure);
             }
           }
         }
       }
+
       for (DependencyNode child : dependencyNode.getChildren()) {
+        // A graph including provided scope may become very big (more than 20,000 nodes).
+        // https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/367
+        // Optimizing the traversal by not visiting a node twice
+        if (includeProvidedScope && !visitedArtifacts.add(child.getArtifact())) {
+          // The optimization is not applicable for graphs used for upper-bound suggestions.
```

Comment:
Why? does a node here not include version?

Reply:
I still l don't follow this. Why would we ever want to traverse a path from the same artifact a second time?

Reply:
It's because a tree starting with each artifact should have complete upper-bound suggestion.

For example, when Artifact A and B show upper-bound suggestions for artifact X, following suggestions are expected:

```
(expected)
Artifact A: version 1
  upper-bound suggestion: artifact X: version 2
Artifact B: version 1
  upper-bound suggestion: artifact X: version 2
```

But the optimization skips artifact X in artifact B, resulting in following suggestions, because artifact C node in artifact B was skipped.

```
(unexpected, missing suggestion for Artifact B because of the optimization)
Artifact A: version 1
  upper-bound suggestion: artifact X: version 2
Artifact B: version 1
  upper-bound suggestion: (none, skipped!)
```

Reply:
Sorry, I'm still not seeing this. Perhaps we should discuss offline with a whiteboard. If the artifact is the same as what we already saw, we should not follow it ever. If the artifact is different than what we saw, we should always follow it. How is anything missed?